latch/registry/table.py

```diff
@@ -0,0 +1,275 @@
+        if not load_if_missing or self._columns is not None:
+            return self._columns
+
+        self._columns = self._get_info()["catalogExperiment"][
+            "catalogExperimentColumnDefinitionsByExperimentId"
+        ]["nodes"]
+
+        return self._columns
+
+    def list_records(self) -> Iterator[ListRecordsOutputType]:
+        has_next_page = True
+        end_cursor = None
+
+        while has_next_page:
+            data = execute(
+                """
+                query tableQuery($argTableId: BigInt!, $argAfter: Cursor) {
+                    catalogExperiment(id: $argTableId) {
+                        catalogSamplesByExperimentId(
+                            condition: { removed: false }
+                            first: 50
+                            after: $argAfter
+                        ) {
+                            nodes {
+                                id
+                                name
+                                catalogSampleColumnDataBySampleId {
+                                    nodes {
+                                        data
+                                        key
+                                    }
+                                }
+                            }
+                            pageInfo {
+                                endCursor
+                                hasNextPage
+                            }
+                        }
+                    }
+                }
+                """,
+                {"argTableId": self.id, "argAfter": end_cursor},
+            )["catalogExperiment"]["catalogSamplesByExperimentId"]
+
+            records = data["nodes"]
+            end_cursor = data["pageInfo"]["endCursor"]
+            has_next_page = data["pageInfo"]["hasNextPage"]
+
+            output: List[Record] = []
+            errors: List[Exception] = []
+
+            for record in records:
+                record_data = record["catalogSampleColumnDataBySampleId"]["nodes"]
+
+                values = {}
+
+                try:
+                    for column in self.get_columns(load_if_missing=True):
+                        key = column["key"]
+                        typ = column["type"]
+
+                        for data_point in record_data:
+                            if data_point["key"] != key:
+                                continue
+                            values[key] = to_python_literal(
+                                data_point["data"],
+                                typ["type"],
+                            )
+                            break
+
+                        if key not in values:
+                            if not typ["allowEmpty"]:
+                                raise ValueError(
+                                    f"no value provided for required column {key}"
+                                )
+                            values[key] = None
+                except ValueError as e:
```

Comment:
I'd maybe `yield` an error immediately, then the user code could either request the rest of the page or abort on the spot

but not sure if that makes for a good API, can leave it as is for now, maybe this will be an option in the future

Reply:
noted, added a todo